File: apps/android/app/src/main/java/com/voiid/app/net/BackupService.kt
package com.voiid.app.net

import android.content.Context
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import okhttp3.OkHttpClient
import okhttp3.Request
import java.util.concurrent.TimeUnit

/**
 * Encrypted-backup blob transport. The blob is sealed ON-DEVICE
 * (encryptBackup under the master secret) before it leaves; the server stores opaque
 * bytes in R2 and never sees plaintext or the key.
 *   - PUT /backup   RAW application/octet-stream body → {stored, size_bytes} (max 50 MiB)
 *   - GET /backup   → {download_url (presigned R2), size_bytes, updated_at}; 404 = none
 *
 * The raw PUT and JSON metadata GET go through [ApiClient.requestRaw] (JWT-attached,
 * versioned /v1/backup). The presigned download is a plain GET to R2 (mirrors
 * [MediaService]'s blobClient).
 */
class BackupService(context: Context) {
    private val api = ApiClient(TokenStore.get(context))

    private val blobClient = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .writeTimeout(120, TimeUnit.SECONDS)
        .readTimeout(120, TimeUnit.SECONDS)
        .build()

    @Serializable
    data class BackupMeta(
        val download_url: String? = null,
        val size_bytes: Long = 0,
        val updated_at: String? = null,
    )

    @Serializable private data class UploadResp(val stored: Boolean = false, val size_bytes: Long = 0)

    /** Upload the encrypted backup blob (raw octet-stream). */
    suspend fun uploadBackup(blob: ByteArray) {
        val resp = api.requestRaw("PUT", "backup", body = blob, contentType = "application/octet-stream")
        if (!resp.isSuccessful) {
            throw ApiError.Http(resp.code, "Backup upload failed (${resp.code}).")
        }
    }

    /** Backup metadata (last size / time + presigned download url), or null if none (404). */
    suspend fun fetchBackupMeta(): BackupMeta? {
        val resp = api.requestRaw("GET", "backup")
        return when {
            resp.code == 404 -> null
            resp.isSuccessful -> ApiClient.json.decodeFromString(BackupMeta.serializer(), String(resp.body))
            else -> throw ApiError.Http(resp.code, "Couldn't fetch backup info (${resp.code}).")
        }
    }

    /** Download the encrypted backup blob via the presigned R2 url. */
    suspend fun downloadBackup(): ByteArray {
        val meta = fetchBackupMeta() ?: throw ApiError.Http(404, "No backup found.")
        val url = meta.download_url ?: throw ApiError.Http(404, "No backup found.")
        return withContext(Dispatchers.IO) {
            val req = Request.Builder().url(url).get().build()
            blobClient.newCall(req).execute().use {
                if (!it.isSuccessful) throw ApiError.Http(it.code, "Backup download failed (${it.code}).")
                it.body?.bytes() ?: ByteArray(0)
            }
        }
    }
}

File: apps/ios/Voiid/Voiid/Networking/BackupService.swift
//
//  BackupService.swift
//  Voiid
//
//  Transport for the encrypted message-history backup blob. The blob is sealed
//  on-device (`encryptBackup`, e2e-core) under the backup master secret before it
//  ever leaves — the server stores opaque ciphertext in R2 and never sees the key
//  or the plaintext. This service mirrors MediaService's raw-URLSession pattern for
//  the octet-stream PUT and the presigned R2 download; the metadata GET goes through
//  the JSON APIClient.
//
//    GET /v1/backup  → { download_url, size_bytes, updated_at }  | 404 no backup
//    PUT /v1/backup  (application/octet-stream, ≤ 50 MiB)        → { stored, size_bytes }
//    download_url    plain GET from R2 (presigned, unauthenticated)
//

import Foundation

/// Backup metadata returned by GET /backup. `download_url` is a short-lived presigned
/// R2 URL fetched with a plain (unauthenticated) GET.
struct BackupMeta: Decodable, Identifiable {
    let download_url: String
    let size_bytes: Int
    let updated_at: String

    /// Identity for SwiftUI `.sheet/.fullScreenCover(item:)`. The presigned URL is
    /// unique per fetch; that's fine for transient presentation.
    var id: String { download_url }

    /// `updated_at` parsed as a Date (ISO-8601), for display.
    var updatedAtDate: Date? { ISO8601DateFormatter().date(from: updated_at) }
}

@MainActor
final class BackupService {
    static let shared = BackupService()
    private let api = APIClient()
    private init() {}

    /// Server-side cap. We reject oversized blobs client-side with a clear message
    /// rather than letting the PUT fail opaquely.
    static let maxBytes = 50 * 1024 * 1024

    /// Fetch backup metadata, or nil when there is no backup yet (404).
    func fetchBackupMeta() async throws -> BackupMeta? {
        do { return try await api.request("GET", "backup") as BackupMeta }
        catch APIError.http(let status, _) where status == 404 { return nil }
    }

    /// Raw octet-stream upload of the sealed blob. Mirrors MediaService: manual
    /// bearer auth + URLSession (the JSON APIClient can't send a raw body).
    func uploadBackup(_ blob: Data) async throws {
        guard blob.count <= Self.maxBytes else {
            throw APIError.http(status: 413, message: "Backup is too large to upload.")
        }
        guard let token = TokenStore.shared.jwt else { throw APIError.notAuthenticated }
        let base = APIConfig.baseURL.absoluteString
        let full = (base.hasSuffix("/") ? base : base + "/") + "\(APIConfig.apiVersion)/backup"
        guard let url = URL(string: full) else { throw APIError.http(status: 0, message: "bad backup url") }
        var req = URLRequest(url: url)
        req.httpMethod = "PUT"
        req.setValue("application/octet-stream", forHTTPHeaderField: "Content-Type")
        req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        req.httpBody = blob

        let resp: URLResponse
        do { (_, resp) = try await URLSession.shared.data(for: req) }
        catch { throw APIError.transport(error) }
        let status = (resp as? HTTPURLResponse)?.statusCode ?? 0
        guard (200..<300).contains(status) else {
            throw APIError.http(status: status, message: "Backup upload failed (\(status)).")
        }
    }

    /// Download the sealed blob: GET metadata for the presigned URL, then a plain GET
    /// of the ciphertext from R2. Throws 404 if there is no backup.
    func downloadBackup() async throws -> Data {
        guard let meta = try await fetchBackupMeta() else {
            throw APIError.http(status: 404, message: "No backup to download.")
        }
        guard let url = URL(string: meta.download_url) else {
            throw APIError.http(status: 0, message: "bad download url")
        }
        let data: Data
        let resp: URLResponse
        do { (data, resp) = try await URLSession.shared.data(from: url) }
        catch { throw APIError.transport(error) }
        let status = (resp as? HTTPURLResponse)?.statusCode ?? 0
        guard (200..<300).contains(status) else {
            throw APIError.http(status: status, message: "Backup download failed (\(status)).")
        }
        return data
    }
}
